src/common/tools/semver.cpp

```diff
@@ -0,0 +1,434 @@
+        }
+
+        if (isNumeric2 && !isNumeric1) {
+            return {1, ErrorEnum::eNone};
+        }
+
+        if (isNumeric1 && isNumeric2) {
+            if (auto result = CompareNumericIdentifiers(identifiers1[i], identifiers2[i]);
+                !result.mError.IsNone() || result.mValue != 0) {
+                return result;
+            }
+        }
+
+        if (auto result = CompareStrIdentifiers(identifiers1[i], identifiers2[i]);
+            !result.mError.IsNone() || result.mValue != 0) {
+            return result;
+        }
+
+        continue;
+    }
+
+    if (identifiers1.Size() < identifiers2.Size()) {
+        return {-1, ErrorEnum::eNone};
+    }
+
+    if (identifiers1.Size() > identifiers2.Size()) {
+        return {1, ErrorEnum::eNone};
+    }
+
+    return {0, ErrorEnum::eNone};
+}
+
+} // namespace
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Error ValidateSemver(const String& version)
+{
+    StaticString<cVersionLen> tmpVersion(version);
+    StaticArray<String, 3>    parts;
+
+    if (auto err = SplitVersion(tmpVersion, parts); !err.IsNone()) {
+        return err;
+    }
+
+    if (auto err = ValidateBasePart(parts[0]); !err.IsNone()) {
+        return err;
+    }
+
+    if (!parts[1].IsEmpty()) {
+        if (auto err = ValidatePrereleasePart(parts[1]); !err.IsNone()) {
+            return err;
+        }
+    }
+
+    if (!parts[2].IsEmpty()) {
+        if (auto err = ValidateMetadataPart(parts[2]); !err.IsNone()) {
+            return err;
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+RetWithError<int> CompareSemver(const String& version1, const String& version2)
+{
+    if (auto err = ValidateSemver(version1); !err.IsNone()) {
+        return {0, err};
+    }
+
+    if (auto err = ValidateSemver(version2); !err.IsNone()) {
+        return {0, err};
+    }
+
+    StaticString<cVersionLen> tmpVersion1(version1);
+    StaticString<cVersionLen> tmpVersion2(version2);
+    StaticArray<String, 3>    parts1;
+    StaticArray<String, 3>    parts2;
+
+    if (auto err = SplitVersion(tmpVersion1, parts1); !err.IsNone()) {
+        return {0, err};
+    }
+
+    if (auto err = SplitVersion(tmpVersion2, parts2); !err.IsNone()) {
+        return {0, err};
+    }
+
+    if (auto result = CompareBaseParts(parts1[0], parts2[0]); !result.mError.IsNone() || result.mValue != 0) {
+        return result;
+    }
+
+    if (!parts1[1].IsEmpty() && parts2[1].IsEmpty()) {
+        return {-1, ErrorEnum::eNone};
+    }
+
+    if (parts1[1].IsEmpty() && !parts2[1].IsEmpty()) {
+        return {1, ErrorEnum::eNone};
+    }
+
+    if (parts1[1].IsEmpty() && parts2[1].IsEmpty()) {
+        return {0, ErrorEnum::eNone};
+    }
+
+    return ComparePrereleaseParts(parts1[1], parts2[1]);
```

Comment:
metadata part isn't checked

Reply:
metadata isn't taken into account for semver comparison according to spec